library/std/src/sys/alloc/vexos.rs

```diff
@@ -0,0 +1,98 @@
+    fn alloc(&self, _size: usize) -> (*mut u8, usize, u32) {
+        static INIT: AtomicBool = AtomicBool::new(false);
+
+        if !INIT.swap(true, Ordering::Relaxed) {
+            unsafe {
+                (
+                    ptr::addr_of_mut!(__heap_start).cast(),
+                    ptr::addr_of!(__heap_end).byte_offset_from(ptr::addr_of!(__heap_start)) as _,
+                    0,
+                )
+            }
+        } else {
+            (ptr::null_mut(), 0, 0)
+        }
+    }
+
+    fn remap(&self, _ptr: *mut u8, _oldsize: usize, _newsize: usize, _can_move: bool) -> *mut u8 {
+        ptr::null_mut()
+    }
+
+    fn free_part(&self, _ptr: *mut u8, _oldsize: usize, _newsize: usize) -> bool {
+        false
+    }
+
+    fn free(&self, _ptr: *mut u8, _size: usize) -> bool {
+        false
+    }
+
+    fn can_release_part(&self, _flags: u32) -> bool {
+        false
+    }
+
+    fn allocates_zeros(&self) -> bool {
+        false
+    }
+
+    fn page_size(&self) -> usize {
+        0x1000
+    }
+}
+
+#[stable(feature = "alloc_system_type", since = "1.28.0")]
+unsafe impl GlobalAlloc for System {
+    #[inline]
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        // SAFETY: DLMALLOC access is guaranteed to be safe because the lock gives us unique and non-reentrant access.
+        // Calling malloc() is safe because preconditions on this function match the trait method preconditions.
+        let _lock = lock::lock();
+        unsafe { DLMALLOC.malloc(layout.size(), layout.align()) }
+    }
+
+    #[inline]
+    unsafe fn alloc_zeroed(&self, layout: Layout) -> *mut u8 {
+        // SAFETY: DLMALLOC access is guaranteed to be safe because the lock gives us unique and non-reentrant access.
+        // Calling calloc() is safe because preconditions on this function match the trait method preconditions.
+        let _lock = lock::lock();
+        unsafe { DLMALLOC.calloc(layout.size(), layout.align()) }
+    }
+
+    #[inline]
+    unsafe fn dealloc(&self, ptr: *mut u8, layout: Layout) {
+        // SAFETY: DLMALLOC access is guaranteed to be safe because the lock gives us unique and non-reentrant access.
+        // Calling free() is safe because preconditions on this function match the trait method preconditions.
+        let _lock = lock::lock();
+        unsafe { DLMALLOC.free(ptr, layout.size(), layout.align()) }
+    }
+
+    #[inline]
+    unsafe fn realloc(&self, ptr: *mut u8, layout: Layout, new_size: usize) -> *mut u8 {
+        // SAFETY: DLMALLOC access is guaranteed to be safe because the lock gives us unique and non-reentrant access.
+        // Calling realloc() is safe because preconditions on this function match the trait method preconditions.
+        let _lock = lock::lock();
+        unsafe { DLMALLOC.realloc(ptr, layout.size(), layout.align(), new_size) }
+    }
+}
+
+mod lock {
+    #[inline]
+    pub fn lock() {} // we don't have threads, which makes this real easy
+}
```

Comment:
Are there no asynchronous events, like interrupts, that would make attempting to access the allocator in the middle of another attempt to access the allocator possible?

Reply:
No, there is no way for user code to observe any concurrency without using unsafe code to manually install interrupt handlers in the boot vector table.

Reply:
I see. Then that code would have to assume that the Rust runtime is not online yet, since it's part of the boot vector table. Do we expect this to be normally done? I'm not sure how easy it is to program these sorts of details for VEX specifically.

Reply:
I'm not aware of any specific situations where users would want this, but we could just wrap allocation in a critical section out of an abundance of caution. I expect that 99.99% of users will use this target with either vexide, which intentionally avoids any preemptive concurrency, or vex-rt, which would use `no_std` anyway, but the cost of adding a critical section is quite low so it's probably a good idea anyway.

Reply:
> I see. Then that code would have to assume that the Rust runtime is not online yet, since it's part of the boot vector table. Do we expect this to be normally done? I'm not sure how easy it is to program these sorts of details for VEX specifically.

I'm not sure if this is particularly relevant to allocation locks, but the Rust runtime is not part of the boot vector table on this target.

Just to clarify a bit about user program startup - VEXos sets up the "userspace" (a block of memory starting at `0x03800000`) and begins executing instructions at `0x03800020`. No interrupt vectors or any kind of real "boot code" is needed, as the user processor is already fully setup with its own interrupt handlers by the time any Rust code is executed. FPU, interrupt handlers, and timer peripherals are already setup for us. All that VEXos needs as an entrypoint is some instructions at `0x03800020` which in our case is some `global_asm` we toss into our linkerscript's `.boot` section that just loads the stack pointer before branching into actual Rust code.

The only time Rust code would interact with interrupts is if the user explicitly decided to overwrite VEXos's [VBAR](https://developer.arm.com/documentation/ddi0601/latest/AArch32-Registers/VBAR--Vector-Base-Address-Register) with its own vector table, which probably is a valid case for wrapping allocation in a critical section.

Reply:
exciting!

yeah, I was mostly just asking to get a sense of where it was on the sliding scale between "full-fledged preemptive multitasking OS" and "barely more than firmware".

Reply:
> I'm not sure if this is particularly relevant to allocation locks, but the Rust runtime is not part of the boot vector table on this target.

Oh and also just to be clear I am aware of this, I just meant that "code in the boot vector cannot assume a runtime that won't appear until well after boot is active". :3